File: examples/position.cpp
#include <iostream>

#include <ruckig/ruckig.hpp>


using namespace ruckig;

int main() {
    // Create instances: the ruckig otg as well as input and output parameters
    Ruckig<6> otg {0.001};
    InputParameter<6> input;
    OutputParameter<6> output;

    // Set input parameters
    input.max_velocity = {1.2, 1.2, 1.2, 0.6, 0.6, 0.6};
    input.max_acceleration = {4.0, 4.0, 4.0, 1.5, 1.5, 1.5};
    input.max_jerk = {10.0, 10.0, 10.0, 4.0, 4.0, 4.0};

    input.current_position = {0.0, -0.1, 0.12, 0.0, 0.3, 0.05};
    input.current_velocity = {0.0, 0.0, 0.2, 0.0, 0.0, 0.0};
    input.current_acceleration = {0.0, 0.0, 0.0, 0.0, 0.0, 0.0};

    input.target_position = {1.0, 0.5, 0.5, 0.0, -0.1, 0.2};
    input.target_velocity = {0.0, 0.0, 0.0, 0.0, 0.0, 0.0};
    input.target_acceleration = {0.0, 0.0, 0.0, 0.0, 0.0, 0.0};

    // Generate trajectory
    std::cout << "t | p1 | p2 | p3 | p4 | p5 | p6" << std::endl;
    while (otg.update(input, output) == Result::Working) {
        auto& p = output.new_position;
        std::cout << output.time << " " << p[0] << " " << p[1] << " " << p[2] << " " << p[3] << " " << p[4] << " " << p[5] << std::endl;

        input.current_position = output.new_position;
        input.current_velocity = output.new_velocity;
        input.current_acceleration = output.new_acceleration;
    }

    std::cout << "Reached target position in " << output.trajectory.get_duration() << " [s]." << std::endl;
}

File: examples/position.py
from copy import copy
from pathlib import Path
from sys import path

# Path to the build directory including a file similar to 'ruckig.cpython-37m-x86_64-linux-gnu'.
build_path = Path(__file__).parent.parent / 'build'
path.insert(0, str(build_path / 'Debug'))  # For windows
path.insert(0, str(build_path))

from ruckig import InputParameter, OutputParameter, Result, Ruckig


if __name__ == '__main__':
    inp = InputParameter(3)
    inp.current_position = [0.2, 0, -1]
    inp.current_velocity = [0, 0.2, 0]
    inp.current_acceleration = [0, 1, 0]
    inp.target_position = [0, -1, -1]
    inp.target_velocity = [0.2, 0, 0]
    inp.target_acceleration = [0, 0.1, -0.1]
    inp.max_velocity = [2, 1, 1]
    inp.max_acceleration = [0.2, 2, 2]
    inp.max_jerk = [3, 4, 5]

    otg = Ruckig(3, 0.05)

    out = OutputParameter(3)
    first_output = None

    print('\t'.join(['t'] + [str(i) for i in range(otg.degrees_of_freedom)]))

    res = Result.Working
    while res == Result.Working:
        res = otg.update(inp, out)

        print('\t'.join([f'{out.time:0.3f}'] + [f'{p:0.3f}' for p in out.new_position]))

        inp.current_position = out.new_position
        inp.current_velocity = out.new_velocity
        inp.current_acceleration = out.new_acceleration

        if not first_output:
            first_output = copy(out)

    print(f'Calculation duration: {first_output.calculation_duration:0.1f} [µs]')
    print(f'Trajectory duration: {first_output.trajectory.duration:0.4f} [s]')
